### ISSUE TITLE
Support for managing individual key versions in google_kms_crypto_key_version 

### ISSUE THREAD

**Issue by toyopilgrim** [hashicorp/terraform-provider-google#20298]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to a user, that user is claiming responsibility for the issue.
* Customers working with a Google Technical Account Manager or Customer Engineer can ask them to [reach out internally](https://github.com/hashicorp/terraform-provider-google/wiki/Customer-Contact#raising-gcp-internal-issues-with-the-provider-development-team) to expedite investigation and resolution of this issue.

### Description

Currently, google_kms_crypto_key_version resource lacks the ability to specify a particular key version and control its state (e.g., setting DESTROY_SCHEDULED, DISABLED). This limits Terraform users’ ability to manage key lifecycle processes, including rotation, disabling, and destruction of specific key versions within Google Cloud KMS.

The feature request is to enhance the google_kms_crypto_key_version resource to support specifying an individual key version and allowing state management actions. 

### New or Affected Resource(s)

* google_kms_crypto_key_version

### Potential Terraform Configuration

```tf 
resource "google_kms_crypto_key_version" "example-key-v1-deletion" {
  crypto_key = google_kms_crypto_key.cryptokey.id
  version = 1
  action = "DESTROY"
}
```

**Comment by BBBmau** [comment]
Based off of the API https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings.cryptoKeys.cryptoKeyVersions there is no `action` field and it also seems like it's not expected to be able to control the state since it's expected that the state will automatically change based on the actions from Cloud KMS and the configuration. Could you provide a link that shows the ability to set the action for a cryptoKeyVersion?

**Comment by toyopilgrim** [comment]
Thank you for the response. You are correct that the API does not include `action` field.  But `state` can be controlled with the following methods:
https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings.cryptoKeys.cryptoKeyVersions/patch
https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings.cryptoKeys.cryptoKeyVersions/destroy#google.cloud.kms.v1.KeyManagementService.DestroyCryptoKeyVersion

The alternative field for action would be like this. Does it make sense?
```
resource "google_kms_crypto_key_version" "example" {
  crypto_key = google_kms_crypto_key.cryptokey.id
  version    = 1
  state      = "DESTROY_SCHEDULED" # Maps to the "destroy" API method
}
```

**Comment by toyopilgrim** [comment]
@BBBmau  Forgot to mention. Just a reminder.

**Comment by tdbhacks** [comment]
Apologies if I'm misunderstanding here, I haven't used Terraform very much: if I understand correctly, each TF `google_kms_crypto_key_version` resource in your TF config maps to a KMS CryptoKeyVersion (which gets initially created on `terraform apply`). So each resource always refers to a single CKV, it's wouldn't make sense to have a `version` field in there because it already points to a single version.

When the CKV is created, doesn't the existing [`state` field](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/kms_crypto_key_version#state-1) allow you to update each CKV's state individually? When set/updated, I was imagining that TF would be sending a :patch request to KMS and allow you to update a key to eg. `DISABLED`. And on destroy, TF calls :destroycryptokeyversion and causes the CKV state to automatically be updated to `DESTROYED`

Is this not how it's actually behaving?

**Comment by toyopilgrim** [comment]
@tdbhacks 
Thank you for taking a look at the issue! 
You are right about that the existing key version can be identified by the terraform original resource reference. So state works for the target as intended. I agree with that having a version field would not make sense.

One remaining issue for me is that neither DESTROYED nor DESTROY_SCHEDULED does not work with the following error : `Error 400: Value DESTROY_SCHEDULED for field crypto_key_version.state is invalid for this request.` 
Both `DISABLED` and `ENABLED` work fine whereas.

**Comment by tdbhacks** [comment]
@toyopilgrim ah yes, so for `DESTROYED` and `DESTROY_SCHEDULED`, you wouldn't call UpdateCryptoKeyVersion (`:patch`) on the CryptoKeyVersion. Those states are only used as part of the `DestroyCryptoKeyVersion` flow.

In Terraform, this means that when you remove the `google_kms_crypto_key_version` from your TF config (or run `terraform destroy` I believe, someone on the terraform team should be able to confirm), the CryptoKeyVersion is automatically scheduled for destruction and transitions to `DESTROY_SCHEDULED` state. Later on, when the configured `destroy_scheduled_duration` time has passed (by default, CKVs stay in `DESTROY_SCHEDULED` state for 30 days), the CKV automatically transitions to `DESTROYED` state.

That said, one thing I'm not sure is working in Terraform or not, is the ability to restore a CKV. Normally, you can restore a CKV while it's in `DESTROY_SCHEDULED` state, if you want to undo the destruction. But I'm not sure if adding the `google_kms_crypto_key_version` back in the config actually works like that, it might not.

Hope this helps :)

**Comment by toyopilgrim** [comment]
@tdbhacks 
Yes, removing the config allows a key version to become `DESTROY_SCHEDULED` as you mentioned.  
Once it's removed, adding the `google_kms_crypto_key_version` back in the config just creates a new key version.  (I guess that it's an expected behavior.)
Hence there currently seems no way to restore a particular key version in Terraform? 

If `state` field add support for `DESTROY_SCHEDULED`, resetting `ENABLED/DISABLED` might be the equivalent to restoring. 

> [state](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/kms_crypto_key_version#state-2) - (Optional) The current state of the CryptoKeyVersion. Possible values are: PENDING_GENERATION, ENABLED, DISABLED, DESTROYED, DESTROY_SCHEDULED, PENDING_IMPORT, IMPORT_FAILED.

At the moment, this document might be confusing because there are not working values described such as `DESTROY_SCHEDULED` and `DESTROY`.

**Comment by tdbhacks** [comment]
@toyopilgrim thanks for taking the time to test this, much appreciated! Indeed, that would have been my guess, it might be that restore is just too complex to support in Terraform or that the TF maintainers have identified other concerns.

@BBBmau do you happen to know if restoring Cloud KMS CryptoKeyVersions was ever considered as a functionality to be supported in TF? Feel free to loop in anyone else you think may have context.

> At the moment, this document might be confusing because there are not working values described such as `DESTROY_SCHEDULED` and `DESTROY`.

I agree that as-is, the documentation is a bit confusing, I'll send a quick PR shortly to try to rephrase. Thank you!
